pkg/resources/secret_with_basic_authentication.go

```diff
@@ -0,0 +1,163 @@
+package resources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/helpers"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/internal/provider"
+	"github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+var secretBasicAuthenticationSchema = func() map[string]*schema.Schema {
+	secretBasicAuthentication := map[string]*schema.Schema{
+		"username": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Specifies the username value to store in the secret when setting the TYPE value to PASSWORD.",
+		},
+		"password": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Specifies the password value to store in the secret when setting the TYPE value to PASSWORD.",
+		},
+	}
+	return helpers.MergeMaps(secretCommonSchema, secretBasicAuthentication)
+}()
+
+func SecretWithBasicAuthentication() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: CreateContextSecretWithBasicAuthentication,
+		ReadContext:   ReadContextSecretWithBasicAuthentication,
+		UpdateContext: UpdateContextSecretWithBasicAuthentication,
+		DeleteContext: DeleteContextSecretWithBasicAuthentication,
+
+		Schema: secretBasicAuthenticationSchema,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Description: "Secret with Basic Authentication where Secrets Type attribute is set to PASSWORD.",
+	}
+}
+
+func CreateContextSecretWithBasicAuthentication(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*provider.Context).Client
+	commonCreate := handleSecretCreate(d)
+
+	id := sdk.NewSchemaObjectIdentifier(commonCreate.database, commonCreate.schema, commonCreate.name)
+
+	usernameString := d.Get("username").(string)
+	passwordString := d.Get("password").(string)
+
+	request := sdk.NewCreateWithBasicAuthenticationSecretRequest(id, usernameString, passwordString)
+	if v, ok := d.GetOk("comment"); ok {
+		request.WithComment(v.(string))
+	}
+
+	err := client.Secrets.CreateWithBasicAuthentication(ctx, request)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(helpers.EncodeResourceIdentifier(id))
+
+	return ReadContextSecretWithBasicAuthentication(ctx, d, meta)
+}
+
+func ReadContextSecretWithBasicAuthentication(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*provider.Context).Client
+	id, err := sdk.ParseSchemaObjectIdentifier(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	secret, err := client.Secrets.ShowByID(ctx, id)
+	if secret == nil || err != nil {
```

Comment:
`secret == nil` is rather expected when `err != nil` and we usually not check it. `secret == nil` when `err == nil` would mean there's a bug in ShowByID, but it would be detected rather quickly.

Reply:
refactored